apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java

```diff
@@ -136,9 +141,38 @@ private void reportDeviceInfo(Collection<AudioDeviceInfo> devices) {
             report.append(item);
         }
         report.append(reportAllMicrophones());
+        report.append(reportUsbDevices());
         log(report.toString());
     }
 
+    public String reportUsbDevices() {
+        StringBuffer report = new StringBuffer();
+        report.append("\n############################");
+        report.append("\nUsb Device Report:\n");
+        try {
+            HashMap<String, UsbDevice> usbDeviceList = mUsbManager.getDeviceList();
+            for (UsbDevice usbDevice : usbDeviceList.values()) {
+                report.append("\n==== USB Device ========= " + usbDevice.getDeviceId());
+                report.append("\nProduct Name       : " + usbDevice.getProductName());
+                report.append("\nProduct ID         : 0x" + Integer.toHexString(usbDevice.getProductId()));
+                report.append("\nManufacturer Name  : " + usbDevice.getManufacturerName());
+                report.append("\nVendor ID          : 0x" + Integer.toHexString(usbDevice.getVendorId()));
+                report.append("\nDevice Name        : " + usbDevice.getDeviceName());
+                report.append("\nDevice Protocol    : " + usbDevice.getDeviceProtocol());
+                report.append("\nDevice Class       : " + usbDevice.getDeviceClass());
+                report.append("\nDevice Subclass    : " + usbDevice.getDeviceSubclass());
+                report.append("\nVersion            : " + usbDevice.getVersion());
+                report.append("\n" + usbDevice);
```

Comment:
This is still dumping the massive toString() result. That makes it very hard to read the report.
Is there anything critical in the full dump?

Reply:
See UsbDevice.toString() here: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;bpv=1;bpt=1;l=327?q=usbdevice

Some fields like [mHasAudioPlayback](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;l=65?q=usbdevice) and [mHasMidi](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;bpv=1;bpt=1;l=67?q=usbdevice&gsn=mHasMidi&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%2Fmain%2F%2Fmain%3Flang%3Djava%3Fpath%3Dandroid.hardware.usb.UsbDevice%236a70250472e668eb6e9eccb9550691d95a06d1b8b8dcee3e12c64c123b26ca1d) are only exposed through toString()